### ISSUE TITLE
Some Links stall after finishing scrape; progress stuck at 0% or 1%

### ISSUE THREAD

**Issue by nbdurham** [Jules-WinnfieldX/CyberDropDownloader#324]
I am having an issue with several links. The links get scraped but then progress halts and download never completes. Transfer rate is a “?/Mb/s”. I verified that ample space available on disc. For example here are some problematic links I’ve experienced this issue with:

https://cyberdrop.me/a/NvapvZF7
https://cyberdrop.me/a/uh5xI7ng
https://bunkr.su/a/3Cbjbz1O > for this one, until yesterday I was using version 2 release. Using the start windows.bat with that distribution this link was parsed as having no links. However switching to latest release caused this link to stall as the first two. On the previous distribution the two above links showed the same behaviour. 

Thanks in advance for your help !

both links open in browser.

**Comment by Jules-WinnfieldX** [comment]
Hey, I'd need to see the downloader.log file after you try

**Comment by nbdurham** [comment]
Hey thank you for the reply.

I think that I might have accidentally used v2 for the kida vids when I tried recently. That link works now when I retried, but this time I definitely used the v3 dist. However the cyberdrop files still do not load. Ive attached the log file that contains output for three downloads batch runs running in parallel, the first two with one link each in URLs.txt, and the second with the two cyberdrop links. I have tried both links separately and neither work. If you need more from me I will be happy to provide whatever it is.

B

**Comment by Jules-WinnfieldX** [comment]
Hey Durham, email attachments don't work with github. 

But, for Cyberdrop, the issue you are facing is because their certificate has expired and Cyberdrop-DL doesn't allow insecure connections. There will be an option to allow them in an upcoming release 4.0.1

**Comment by nbdurham** [comment]
Can you point me to a reference of some kind so I can understand better what you are talking about? I still see "https://" when I go to that URL in my browser.

**Comment by nbdurham** [comment]
Here is a snippit of the log just in case.

2023-02-08 18:28:09,455:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0034-raSAq1SVlgRVJci5D5eotuAhMUwitfaO.mp4...
2023-02-08 18:28:10,033:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0111-SB1YCUW64TFGwwDJbMtxWv5jWVW8F8zs.mp4...
2023-02-08 18:28:11,674:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0026-xYcBEMHF2FtldotbdPo7XjG8j1h8PypF.mp4...
2023-02-08 18:28:13,067:DEBUG:downloaders:downloaders.py:39:Retrying (2) https://fs-01.cyberdrop.cc/0135-0lXZNTLEnVw4LoHv9aObiccZQWUDBmpn.mp4...
2023-02-08 18:28:15,114:DEBUG:downloaders:downloaders.py:39:Retrying (2) https://fs-01.cyberdrop.cc/0087-gF0IgWibVNb9uqofVM2sFMGJIcdmx9Oz.mp4...
2023-02-08 18:28:15,247:DEBUG:downloaders:downloaders.py:39:Retrying (1) https://fs-01.cyberdrop.cc/0025-o0HDX9VVAXxsaRXHZP3LJWy6PpBfZwVC.mp4...
2023-02-08 18:28:15,784:DEBUG:downloaders:downloaders.py:39:Retrying (2) https://fs-01.cyberdrop.cc/0028-MusRyzANkybYRYIxg024dCHLB4ryoEe1.mp4...
2023-02-08 18:28:15,887:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0067-07bNnhxsBNSNFIdSw3HgmbJAlkXMRhjK.mp4...
2023-02-08 18:28:17,294:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0095-6efDznKsrTQM4FJ2g9TrZbA1AUHj8RgO.mp4...
2023-02-08 18:28:17,473:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0116-G59Fl1TtCbq8ibOs2WFErwQXsoELrWV0.mp4...
2023-02-08 18:28:17,473:DEBUG:downloaders:downloaders.py:39:Retrying (2) https://fs-01.cyberdrop.cc/0085-Kwe2E5fhuoqiVe1luMzWdtaQGktTESkd.mp4...
2023-02-08 18:28:18,596:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0068-jO8XTbJXdzRq6lZq1Lqg1a8FjDffJXb9.mp4...
2023-02-08 18:28:19,581:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0042-GLqs99Hete8UKRFuKgyErmLSHX041mdk.mp4...
2023-02-08 18:28:20,004:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0090-jcGXRwSDuCuKbR8kQJtDXY5WZKRy1pSx.mp4...
2023-02-08 18:28:20,107:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0045-8CchTeMnnZMCLAw0bYduZSmyQXEFC4Fl.mp4...
2023-02-08 18:28:20,981:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0167-QwU9jRqvHoMv8drogZBsY6ZPcV3ca1hD.mp4...
2023-02-08 18:28:21,279:DEBUG:downloaders:downloaders.py:39:Retrying (2) https://fs-01.cyberdrop.cc/0057-38GQTBGW4AvkOvfIZgliW0wht7elmob1.mp4...
2023-02-08 18:28:21,798:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0126-3UWkosQWrm8pxpbKsumHzKlOydb9X7ze.mp4...
2023-02-08 18:28:22,020:DEBUG:downloaders:downloaders.py:39:Retrying (1) https://fs-01.cyberdrop.cc/0118-TSLAELnapVgyf3BYv4GPv1FWlUr4RcUR.mp4...
2023-02-08 18:28:23,077:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0107-XbhLttJdNd7xBrSe3gKpUBKI96ZldAtT.mp4...
2023-02-08 18:28:23,543:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0069-x7QO7bhxTvxHIM3dmXBd3C06DskOUZi3.mp4...
2023-02-08 18:28:23,960:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0105-AQQtMo6H7sZhqAVdNY59046Vi7kMRg4h.mp4...
2023-02-08 18:28:25,066:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0066-Z7NNP2V6SOWQB9Ma56D4bvyQMcX2MeH7.mp4...
2023-02-08 18:28:26,543:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0174-is9CNqvmCranKHUkt1bQOyfMPTFIgf7Y.mp4...
2023-02-08 18:28:26,659:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0185-z9U0Pp918iR5imMHZRh8N84W9NvcAvJq.mp4...
2023-02-08 18:28:27,080:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0161-dBCk7kelSqMMLGAPijCUyA8jDLpT7cEn.mp4...
2023-02-08 18:28:27,962:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0096-tbk0ekVFdzzhVVf2QSTrkW9zlIyfpSa4.mp4...
2023-02-08 18:28:28,482:DEBUG:downloaders:downloaders.py:39:Retrying (2) https://fs-01.cyberdrop.cc/0117-vnG4Af4eK5RkQhSpdftlarCHGLP8tGgm.mp4...
2023-02-08 18:28:29,472:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0075-xjVGqTfBlBlPaNsq4iht4s9NXDIUYA3L.mp4...
2023-02-08 18:28:30,009:DEBUG:downloaders:downloaders.py:39:Retrying (2) https://fs-01.cyberdrop.cc/0155-vqBegqfNvYyB4AokLjgd8LbqAgRtBEgx.mp4...
2023-02-08 18:28:30,510:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0093-Tu2X3xJP31Ymp63w38DybW6HSElN7rb0.mp4...
2023-02-08 18:28:31,028:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0127-vJfW7XvR05q74IJj8ZZ6bsFZ7v2JV33b.mp4...
2023-02-08 18:28:31,716:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0039-6QvN5WZy9Z99jGzXT5HRiRBAX0BBzJh8.mp4...
2023-02-08 18:28:32,236:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0120-iYoNP5QrdHaQBhqIUM4Kc66NCgaoV8ba.mp4...
2023-02-08 18:28:32,862:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0130-bkbPcrwkpalMg2OyDJZL6F6kHG5zZRxA.mp4...
2023-02-08 18:28:33,589:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0040-b9VskpgsPTVdie4p3OiUPqWIPQbHF7oR.mp4...
2023-02-08 18:28:34,605:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0083-yy9fBx10Ld7kNYSVj3GhTlEZES3taSjP.mp4...
2023-02-08 18:28:35,146:DEBUG:downloaders:downloaders.py:39:Retrying (3) https://fs-01.cyberdrop.cc/0011-geKSAWTPDWI1NaZDbitFo4T7HPuducvc.mp4...
2023-02-08 18:28:35,776:DEBUG:downloaders:downloaders.py:39:Retrying (4) https://fs-01.cyberdrop.cc/0121-kebylWCTsiaXXYWosTrvZTg3KxXDpKZm.mp4...

**Comment by Jules-WinnfieldX** [comment]
![image](https://user-images.githubusercontent.com/61347133/217702661-f3ec2cce-4dd3-496d-bded-01cbece8be13.png)

![image](https://user-images.githubusercontent.com/61347133/217702752-9d002a2b-1d4b-4f90-aacf-f6032b308d25.png)

![image](https://user-images.githubusercontent.com/61347133/217702826-b3078517-93c5-4ea5-b018-4ebe29bbb1a0.png)

The important part of the log is what happens after the retries, I'd give you the excerpt if I had the entire downloader.log of yours.

**Comment by nbdurham** [comment]
[downloader.log](https://github.com/Jules-WinnfieldX/CyberDropDownloader/files/10692971/downloader.log)
Okay here you are

**Comment by nbdurham** [comment]
Its worth mentioning that several files did actually complete but its showing a very abnormal DL rate.
![image](https://user-images.githubusercontent.com/8071625/217703346-248b0bf0-05da-41a7-b841-9d3e5959994c.png)

**Comment by notjustanygeek** [comment]
Would the insecure connections also cause this strange delay?

**Comment by Jules-WinnfieldX** [comment]
Not all of cyberdrops certificates have expired, as you pointed out. Just the main ones to _some_ of their file servers as I showed. 

Cyberdrop-DL won't allow connections that are insecure. The delay as you are putting it is likely the program trying to connect but failing, and retrying for nearly 100 links, by default each link is retried 10 times. There is a large portion of that log missing, I'm running it myself to get the actual error message.

**Comment by notjustanygeek** [comment]
It was still running, but it finished and now I see what you mean. 

```
https://fs-01.cyberdrop.cc/0090-jcGXRwSDuCuKbR8kQJtDXY5WZKRy1pSx.mp4 (Cannot connect to host fs-01.cyberdrop.cc:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: certificate has expired (_ssl.c:1129)')])
2023-02-08 18:37:33,632:DEBUG:downloaders:downloaders.py:344:Error status code: 1
2023-02-08 18:37:34,167:DEBUG:downloaders:downloaders.py:39:Retrying (8) https://fs-01.cyberdrop.cc/0018-FAbLTAiDD1TpFei2mw3uvlM2tcNwEeAo.mp4...
2023-02-08 18:37:34,712:DEBUG:downloaders:downloaders.py:39:Retrying (8) https://fs-01.cyberdrop.cc/0164-utXR9GLuWlbWUY7Dm8MM1wx6jq2AdpcD.mp4...
2023-02-08 18:37:36,865:DEBUG:downloaders:downloaders.py:39:Retrying (5) https://fs-01.cyberdrop.cc/0087-gF0IgWibVNb9uqofVM2sFMGJIcdmx9Oz.mp4...
2023-02-08 18:37:37,213:DEBUG:downloaders:downloaders.py:39:Retrying (8) https://fs-01.cyberdrop.cc/0126-3UWkosQWrm8pxpbKsumHzKlOydb9X7ze.mp4...
2023-02-08 18:37:37,539:DEBUG:downloaders:downloaders.py:39:Retrying (8) https://fs-01.cyberdrop.cc/0085-Kwe2E5fhuoqiVe1luMzWdtaQGktTESkd.mp4...
2023-02-08 18:37:37,855:DEBUG:downloaders:downloaders.py:39:Retrying (8) https://fs-01.cyberdrop.cc/0006-YY88qo9SwxL66XUwZ1k91boQo7qtsIS7.mp4...
2023-02-08 18:37:38,030:DEBUG:downloaders:downloaders.py:37:Skipping https://fs-01.cyberdrop.cc/0120-iYoNP5QrdHaQBhqIUM4Kc66NCgaoV8ba.mp4...
2023-02-08 18:37:38,030:DEBUG:downloaders:downloaders.py:339:1
2023-02-08 18:37:38,030:DEBUG:base_functions:base_functions.py:61:
Error attempting https://fs-01.cyberdrop.cc/0120-iYoNP5QrdHaQBhqIUM4Kc66NCgaoV8ba.mp4
```

**Comment by notjustanygeek** [comment]
Thank you Jules, I really respect your dedication to maintaining this.

**Comment by Jules-WinnfieldX** [comment]
There should be an error message below that. 

Which is essentially:
`ClientConnectorCertificateError(ConnectionKey(host='fs-01.cyberdrop.cc', port=443, is_ssl=True, ssl=<ssl.SSLContext object at 0x0000023817DA0D40>, proxy=None, proxy_auth=None, proxy_headers_hash=4376497125137007687), SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: certificate has expired (_ssl.c:997)'))`

The important part being `[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: certificate has expired` notating that the servers certificate has expired.

**Comment by notjustanygeek** [comment]
Okay. Is the certificate verification a feature of the python library?

**Comment by Jules-WinnfieldX** [comment]
It's a feature of SSL, which in this case comes from a mix of AIOHTTP and Certifi. 

Boiled down to ELI5
SSL is a part of the connection that requires a handshake to take place. This hand shake is the verification of the servers certificate

**Comment by Jules-WinnfieldX** [comment]
Essentially everything secure uses SSL, your web browser will. Which is shown by the first picture I sent back, firefox saying the site wasn't secure and was a potential security risk. The website failed the handshake.

**Comment by notjustanygeek** [comment]
okay that makes sense. thanks for explaining. Therefore, allowing insecure connections means that some bad actor can spy on the data transfer between me and that server? If its not too much trouble I'd like to understand why I _wouldnt_ want to allow this insecure connection when v4 is released

**Comment by Jules-WinnfieldX** [comment]
In essence, spying on the connection, but the connection, and information (or files) being sent back and forth can also be more easily altered. 

Since it's still https, the connection is still encrypted, but the trust that the connection is valid and not being interfered with, is lost. It also loses the trust that the server is who they say they are, and will be sending what should actually be sent.

**Comment by notjustanygeek** [comment]
Thanks, that helps me. Appreciate you!
